server/src/main/java/org/opensearch/wlm/cancellation/DefaultTaskCancellation.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    private Long convertThresholdIntoLong(ResourceType resourceType, Double resourceThresholdInPercentage) {
+        Long threshold = null;
+        if (resourceType == ResourceType.MEMORY) {
+            // Check if resource usage is breaching the threshold
+            threshold = (long) (resourceThresholdInPercentage * HEAP_SIZE_BYTES);
+        } else if (resourceType == ResourceType.CPU) {
+            // Get the total CPU time of the process in milliseconds
+            long cpuTotalTimeInMillis = ProcessProbe.getInstance().getProcessCpuTotalTime();
+            // Check if resource usage is breaching the threshold
+            threshold = (long) (resourceThresholdInPercentage * cpuTotalTimeInMillis);
+        }
+        return threshold;
+    }
+
+    private Long getResourceUsage(QueryGroup queryGroup, ResourceType resourceType) {
+        if (!queryGroupLevelResourceUsageViews.containsKey(queryGroup.get_id())) {
+            return 0L;
+        }
+        return queryGroupLevelResourceUsageViews.get(queryGroup.get_id()).getResourceUsageData().get(resourceType);
+    }
+
+    private boolean isBreachingThreshold(ResourceType resourceType, Double resourceThresholdInPercentage, long resourceUsage) {
+        if (resourceType == ResourceType.MEMORY) {
+            // Check if resource usage is breaching the threshold
+            return resourceUsage > convertThresholdIntoLong(resourceType, resourceThresholdInPercentage);
+        }
+        // Resource types should be CPU, resourceUsage is in nanoseconds, convert to milliseconds
+        long resourceUsageInMillis = resourceUsage / 1_000_000;
+        // Check if resource usage is breaching the threshold
+        return resourceUsageInMillis > convertThresholdIntoLong(resourceType, resourceThresholdInPercentage);
+    }
+
+    private NodeDuressTrackers setupNodeDuressTracker(Settings settings, ClusterSettings clusterSettings) {
+        NodeDuressSettings nodeDuressSettings = new NodeDuressSettings(settings, clusterSettings);
+        return new NodeDuressTrackers(new EnumMap<>(ResourceType.class) {
+            {
+                put(
+                    ResourceType.CPU,
+                    new NodeDuressTrackers.NodeDuressTracker(
+                        () -> ProcessProbe.getInstance().getProcessCpuPercent() / 100.0 >= nodeDuressSettings.getCpuThreshold(),
+                        nodeDuressSettings::getNumSuccessiveBreaches
+                    )
+                );
+                put(
+                    ResourceType.MEMORY,
+                    new NodeDuressTrackers.NodeDuressTracker(
+                        () -> JvmStats.jvmStats().getMem().getHeapUsedPercent() / 100.0 >= nodeDuressSettings.getHeapThreshold(),
+                        nodeDuressSettings::getNumSuccessiveBreaches
+                    )
+                );
+            }
+        });
+    }
```

Comment:
We should rely on cancellation thresholds for the resource type. We are creating two settings for each resource 
1. Cancellation level 
2. Rejection level